Sources/Brave/Frontend/Sync/SyncWelcomeViewController.swift

```diff
@@ -320,32 +353,150 @@ extension SyncWelcomeViewController: SyncPairControllerDelegate {
+              title: Strings.genericErrorTitle,
+              message: Strings.Sync.syncMaximumDeviceReachedErrorDescription,
+              preferredStyle: .alert)
+            alert.addAction(UIAlertAction(title: Strings.OKString, style: .default) { _ in
+              self.leaveIncompleteSyncChain()
+            })
+            present(alert, animated: true, completion: nil)
+          }
         }
       } else {
-        // Show an alert if the sync hain is deleted
-        let alert = UIAlertController(
+        // Show an alert if the sync chain is deleted
+        // Leave sync chain should be called if there is deleted chain alert
+        // to reset sync and local preferences with observer
+        self.clearSyncChainWithAlert(
           title: Strings.Sync.syncChainAlreadyDeletedAlertTitle,
           message: Strings.Sync.syncChainAlreadyDeletedAlertDescription,
-          preferredStyle: .alert)
-        
-        alert.addAction(UIAlertAction(title: Strings.OKString, style: .default) { _ in
-          // Leave sync chain should be called if there is deleted chain alert
-          // to reset sync and local preferences with observer
-          self.syncAPI.leaveSyncGroup()
-
-          controller.disableNavigationPrevention()
-          self.navigationController?.popViewController(animated: true)
-        })
-        
-        self.present(alert, animated: true, completion: nil)
+          controller: controller)
       }
     }
-
-    // In parallel set code words - request sync and setup complete
-    // should be called on brave-core side
-    syncAPI.joinSyncGroup(codeWords: codeWords, syncProfileService: syncProfileServices)
-    handleSyncSetupFailure()
+  }
+  
+  private func retrieveDeviceLimitLevel() -> (level: SyncDeviceLimitLevel?, error: DeviceRetriavalError?) {
+    let deviceListJSON = syncAPI.getDeviceListJSON()
+    let deviceList = fetchSyncDeviceList(listJSON: deviceListJSON)
+    
+    if let error = deviceList.error {
+      return (nil, error)
+    }
+    
+    guard let devices = deviceList.devices else {
+      return (nil, nil)
+    }
+    
+    var deviceLimitLevel: SyncDeviceLimitLevel?
+    
+    switch devices.count {
+    case 1...4:
+      deviceLimitLevel = .safe
+    case 5...9:
+      deviceLimitLevel = .approvalNeeded
+    case 10...:
+      deviceLimitLevel = .blocked
+    default:
+      Logger.module.error("\(DeviceRetriavalError.deviceNumberError.errorDescription)")
+      return (nil, DeviceRetriavalError.deviceNumberError)
+    }
+    
+    return (deviceLimitLevel, nil)
+  }
+  
+  private func fetchNamesOfDevicesInSyncChain() -> [String]? {
+    let deviceListJSON = syncAPI.getDeviceListJSON()
+    let deviceList = fetchSyncDeviceList(listJSON: deviceListJSON)
+    
+    if deviceList.error != nil {
+      return nil
+    }
+    
+    guard let devices = deviceList.devices else {
+      return nil
+    }
+    
+    return devices.map { $0.name ?? "" }
```

Comment:
Should probably default to `unknown device` instead of empty string because it will show up weird in the UI.

Reply:
Actually I want to keep this as it is, inside the loop I am checking if the name is empty or not I think it is better way to take care of this than showing unknown device which can cause some suspicious message there.